### ISSUE TITLE
How can we speedup sonar-flutter?  🚗💨

### ISSUE THREAD

**Issue by mreichelt** [insideapp-oss/sonar-flutter#46]
I'm currently investigating how we can make the sonar-flutter plugin faster by a large margin.

I read #30 by @EnPython and also found out about the PR #29 by @kuhnroyal, which tries to speed up sonar-flutter by using an existing analyzer report file.

After trying out the PR by @kuhnroyal (btw: awesome work! 🎉 ) and running some sonar checks locally, I found out that this indeed speeds up the Sonar check by a large margin.

As a downside, this requires having some sort of `analysis.txt` file generated by some command in the first place. That is why I tried to change the sonar-flutter code to run `flutter analyze` instead of running `dartanalyzer file1 file2 file3 ...`, and the results for our Flutter project was shockingly fast: it took ~23s instead of ~6m, which was a speedup of ~15x.

Of course, the issues found by `flutter analyze` and `dartanalyzer` can be slightly different - and maybe sonar-flutter projects are used by both Flutter and (pure) Dart developers.

That's why I want to know from you: which implementation would you prefer?

1. Change the default from `dartanalyzer` to `flutter analyze`: This would make analysis faster for everyone (yay), but could potentially break for pure-Dart devs
2. Create a setting enum LEGACY/FLUTTER: People could choose the faster Flutter analysis from Sonar settings - but they have to explicitly opt-in (so some people will never know about this option being available)
3. Add an automated check that checks `pubspec.yaml` if Flutter is available and runs `flutter analyze` for Flutter projects and `dartanalyzer` for all others: That way Flutter devs would benefit instantly, and Dart devs would still get the old analysis
4. Some idea of you :)

I'd gladly contribute a PR with this - but it's good to know which direction you would prefer first :)

**Comment by kuhnroyal** [comment]
If I recall correctly, you can actually pass folders to both command, that would speed them up as well.
I would probably go for some `analyzer_command` setting that user can configure, that would allow for all kinds of different ways to execute the analyzer depending on their setup:

```
dartanalyzer
flutter analyze
fvm flutter analyze
pub run dartanalyzer
```

A smart fallback depending on the `pubspec.yaml` sound good.

**Comment by mreichelt** [comment]
@kuhnroyal Thanks for the feedback! That's a good idea indeed. I have to check, because the dartanalyzer/flutter commands have a different output, so that would probably need two settings (one for the dart analyzer command, the other for the flutter analyzer command). To make it even more complicated, there now is a `dart analyze` command. So there actually are 3 ways to analyze Dart/Flutter projects now. Maybe it's a good idea to ask the Flutter team what's the designated future for all of those.

**Comment by mreichelt** [comment]
Ian from Google responded [on Twitter](https://twitter.com/Hixie/status/1361489802705559552):

> If you're working on Flutter-related code, use the flutter tool. If you're working on pure Dart, use the dart tool. They both use dartanalyzer under the hood but that's an implementation detail.

So I think I'll implement a logic that uses `flutter analyze` when the project is a Flutter one (based on the `pubspec.yaml`), and uses `dart analyze` otherwise. I'll see if I can make those commands configurable.

**Comment by mreichelt** [comment]
Good news: running `flutter analyze` also works on Dart projects now. So we can completely skip the `pubspec.yaml` parsing etc. My new approach: Create a setting (legacy/flutter), with flutter as the new default. That way, everyone gets the speedup, while this also works for Flutter as well as Dart projects. And it is a *whole* lot faster!

**Comment by mreichelt** [comment]
I tried it out locally. The new method took 20s, while the legacy one took 166s. Thats a 8x speedup! And no, that's no Aprils fool :)

@kuhnroyal it would be awesome if you could give some feedback to https://github.com/insideapp-oss/sonar-flutter/pull/47/files?diff=unified&w=1 and if you could try this out if this works for you. So we can guarantee that it works on one more machine other than mine ;)

And I would be interested in how fast Sonar gets now when you run this against your project!

**Comment by kuhnroyal** [comment]
> Good news: running `flutter analyze` also works on Dart projects now.

Yes that works but what if there is no flutter installed, should allow `dart analyze` as well I think.

I have not used Sonar for a while so I can't test atm. But I plan to re-integrate into CI soon.

**Comment by mreichelt** [comment]
> Yes that works but what if there is no flutter installed, should allow `dart analyze` as well I think.

Good point - I didn't think of that. I updated the code and added a new 'dart' setting to run this as well - and I tested it locally, too! It's also pushed to the MR 👍

**Comment by zippy1978** [comment]
HI,
Version 0.3.2 is released with @mreichelt MR.
It works much faster now !

**Comment by mreichelt** [comment]
Closed - version 0.3.2 is now around 10x faster than the previous version for most of our builds. Thanks to @zippy1978 for merging, and thanks to everyone for the code reviews! 👏
